### PR TITLE
docs: add Getting Routing Information

### PR DESCRIPTION
**Pull request by datamweb**
**Description**
I was unable to find clear documentation on how to retrieve the controller and method names for the current route in CodeIgniter 4. After searching through the user guide without success, I had to read the source code to figure out the correct approach. While I'm not entirely sure if this has been documented elsewhere, I believe this addition will be helpful for developers who need to dynamically interact with the controller or method handling the current request.

- Retrieving the Controller and Method Names
- Accessing Active Filters in Router

Screenshot 2024-08-19 180907
Screenshot 2024-08-19 181039

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: user_guide_src/source/incoming/routing/071.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+use Config\Services;
+
+// Get the router instance
+$router = Services::router();
```
Comment (kenjis) [inline user_guide_src/source/incoming/routing/071.php:6]: `service()` is recommended. See #8928

```suggestion
$router = service('router');
```

File: user_guide_src/source/incoming/routing.rst [hunk 2]
```diff
@@ -151,6 +151,18 @@ a simple view:
 
 .. literalinclude:: routing/020.php
 
+Retrieving the Controller and Method Names
+------------------------------------------
+
+In some cases, you might need to determine which controller and method have been triggered by the current HTTP request.
+This can be useful for logging, debugging, or conditional logic based on the active controller method.
+
+CodeIgniter 4 provides a simple way to access the current route's controller and method names using the ``Services::router()`` class. Here is an example:
```
Comment (kenjis) [inline user_guide_src/source/incoming/routing.rst:160]:
```suggestion
CodeIgniter 4 provides a simple way to access the current route's controller and method names using the ``Router`` class. Here is an example:
```

**Comment by kenjis** [comment]
I'm not sure that is the right place for this section.
Screenshot 2024-08-18 11 36 13

**Comment by datamweb** [comment]
> I'm not sure that is the right place for this section.
> Isn't it better to move it to the Controllers page?

Honestly, I’m not entirely sure which section would be the most appropriate. However, let me explain why I initially placed this under the Specifying Route Handlers section within the URI Routing page.

This section focuses on different ways to specify what code should handle a particular route. Since retrieving the controller and method names is directly related to how route handlers are resolved by the framework, I felt that placing it here made sense.

The Specifying Route Handlers section already covers topics such as controllers, closures, and array callables, which are different methods for defining route handlers. Including information on retrieving the controller and method names could add valuable context to this section, as it helps in understanding how route handlers are determined and processed.

That said, I’m open to moving this content to another section. Could you please specify the exact location of the **Controllers page** so I can make the necessary changes?

**Comment by neznaika0** [comment]
Yes. The new section looks better.

I've talked about it, there's no way to get the route details. You need to parse it yourself
https://github.com/codeigniter4/CodeIgniter4/blob/f2e560f90c88cc82964f7b5f44819bd47ee4f323/system/Router/RouteCollection.php#L609

**Comment by kenjis** [comment]
@neznaika0 Why do you want to get route options (name, filters, priority...) ?
Use case?

**Comment by datamweb** [comment]
> * This topic is about Routing (Router). So it is better to be in the Routing page.
> * The existing sections are all for how to set routes. But this is how to get the routing info.

Your reasons are quite fair and have explained the issue well.



> get route options (name, filters, priority...) ?
> Use case?

For example(**filters**), if an authentication filter is not active, you might redirect the user to a login page:

```php
$router = service('router');
$filters = $router->getFilters();

if (!in_array('auth', $filters)) {
    return redirect()->to('/login');
}

```

**Comment by datamweb** [comment]
I would like to add a section called **Accessing Active Filters in Router** to **Getting Routing Information** as well. Do you agree with this addition?

**Comment by kenjis** [comment]
@neznaika0 I still don't understand your use case completely. Where do you put the code like that?

File: user_guide_src/source/incoming/routing.rst [hunk 3]
```diff
@@ -1127,3 +1127,36 @@ You can specify the host in the request URL with the ``--host`` option:
...
     php spark routes --host accounts.example.com
+
+Getting Routing Information
+***************************
+
+In CodeIgniter 4, understanding and managing routing information is crucial for handling HTTP requests effectively.
+This involves retrieving details about the active controller and method, as well as the filters applied to a specific route.
+Below, we explore how to access this routing information to assist in tasks such as logging, debugging, or implementing conditional logic.
+
+Retrieving the Controller and Method Names
```
Comment (kenjis) [inline user_guide_src/source/incoming/routing.rst:1138]: Is this better?
```suggestion
Retrieving the Current Controller/Method Names
```

File: user_guide_src/source/incoming/routing.rst [hunk 4]
```diff
@@ -1127,3 +1127,36 @@ You can specify the host in the request URL with the ``--host`` option:
...
+In some cases, you might need to determine which controller and method have been triggered by the current HTTP request.
+This can be useful for logging, debugging, or conditional logic based on the active controller method.
+
+CodeIgniter 4 provides a simple way to access the current route's controller and method names using the ``Router`` class. Here is an example:
+
+.. literalinclude:: routing/071.php
+
+This functionality is particularly useful when you need to dynamically interact with your controller or log which method is handling a particular request.
+
+Accessing Active Filters for a Route
```
Comment (kenjis) [inline user_guide_src/source/incoming/routing.rst:1150]:
```suggestion
Getting Active Filters for the Current Route
```

File: user_guide_src/source/incoming/routing.rst [hunk 5]
```diff
@@ -1127,3 +1127,36 @@ You can specify the host in the request URL with the ``--host`` option:
...
+
+:doc:`Filters <filters>` are a powerful feature that enables you to perform operations such as authentication, logging, and security checks before or after processing HTTP requests.
+To access the active filters for a specific route, you can use the :php:meth:`CodeIgniter\\Router\\Router::getFilters()` method from the ``Router`` class.
+
+This method returns a list of filters that are currently active for the route being processed:
+
+.. literalinclude:: routing/072.php
+
+.. note:: The ``getFilters()`` method returns only the filters defined for the specific route.
+     It does not include global filters or those specified in configuration files.
```
Comment (kenjis) [inline user_guide_src/source/incoming/routing.rst:1161]: This note is good!

```suggestion
     It does not include global filters or those specified in the **app/Config/Filters.php** file.
```

**Comment by neznaika0** [comment]
Example:
```php
if (! empty($routeName) && $routeName === 'homepage') {
            throw new LogicException('Access control: Found redirect loop');
        }
```
It doesn't matter where - it's more reliable than a URL comparison. I use in my "access control" based on controllers (not as Shield)

**Comment by kenjis** [comment]
I still don't understand why you need such code?
Where do you put it?

**Review by kenjis — APPROVED** [review]
LGTM! Thank you.

**Comment by neznaika0** [comment]
@kenjis It's in the filter now. It doesn't matter. The main reason is not to compare dynamic URLs, but a constant route name.
```php
// not
// if (base_url('main') === 'http://example.com/main') {

// better
if (! empty($routeName) && $routeName === 'homepage') {
    throw new LogicException('Access control: Found redirect loop');
}
```

**Comment by kenjis** [comment]
@neznaika0 Okay, thank you for your explanation.

 In my opinion, it is better that the Request (IncomingRequest) has the current route info including the route name.
But this is an enhancement.

**Pull request merged**